src/lib/comms/sol-oic-server.c

```diff
@@ -560,336 +551,129 @@ create_coap_resource(struct sol_str_slice endpoint)
             res->path[current].len++;
     }
 
-    return res;
-}
-
-SOL_API struct sol_oic_device_definition *
-sol_oic_server_register_definition(struct sol_str_slice endpoint,
-    struct sol_str_slice resource_type_prefix,
-    enum sol_coap_flags flags)
-{
-    struct sol_oic_device_definition *def;
-
-    OIC_SERVER_CHECK(NULL);
-
-    def = sol_oic_server_get_definition(endpoint, resource_type_prefix);
-    if (def)
-        return def;
+    res->get = _sol_oic_resource_type_get;
+    res->put = _sol_oic_resource_type_put;
+    res->post = _sol_oic_resource_type_post;
+    res->delete = _sol_oic_resource_type_delete;
 
-    def = sol_vector_append(&oic_server.device_definitions);
-    SOL_NULL_CHECK(def, NULL);
+    if (resource->flags & SOL_OIC_FLAG_DISCOVERABLE)
+        res->flags |= SOL_COAP_FLAGS_WELL_KNOWN;
 
-    def->resource_type_prefix = resource_type_prefix;
-    def->endpoint = endpoint;
-    def->resource = create_coap_resource(endpoint);
-    SOL_NULL_CHECK_GOTO(def->resource, error);
-    def->resource->flags = flags;
-    def->resource->get = _sol_oic_device_definition_specific_get;
-    def->resource->resource_type = resource_type_prefix;
+    res->iface = sol_str_slice_from_str(resource->iface);
+    res->resource_type = sol_str_slice_from_str(resource->rt);
 
-    sol_vector_init(&def->resource_types, sizeof(struct resource_type_data));
-
-    if (!sol_coap_server_register_resource(oic_server.server, def->resource, def))
-        goto error;
-
-    return def;
-
-error:
-    free(def->resource);
-    (void)sol_vector_del(&oic_server.device_definitions, oic_server.device_definitions.len - 1);
-    return NULL;
+    return res;
 }
 
-static void
-_sol_oic_device_definitions_free_resource_types(struct sol_oic_device_definition *def)
+static char *
+create_endpoint(void)
 {
-    struct resource_type_data *iter;
-    uint16_t idx;
+    char *buffer = NULL;
+    int r;
 
-    SOL_VECTOR_FOREACH_REVERSE_IDX (&def->resource_types, iter, idx) {
-        free(iter->resource_type);
-        /* FIXME: sol_coap_server_unregister_resource(iter->resource) */
-        free(iter->resource);
-    }
-    sol_vector_clear(&def->resource_types);
+    r = asprintf(&buffer, "/sol/%d", oic_server.resources.len);
```

Comment:
this is a problem if resources may be deleted and added, then they will clash. I'd use a `unsigned int32_t` as global and keep increasing it, let it overflow as it's less likely to create problems than the current approach.